### DISCUSSION TITLE
How to set credentials in request header when making api call from client to api.

### DISCUSSION THREAD

**Discussion by yotsaphonSutweha** [bali182/oats#150] [Q&A]
Hi there,

First of all I would like to thank you for this library, it's been very helpful and easy to use when dealing with apis. I have been using oats for a client-side project, React specifically. I'm stuck on how to properties like "credentials" etc. in the request header. I tried to find information in the documentation but no luck, or maybe I was missing something. Really appreciate for your help!

**Comment by bali182** [comment]
Hello there, sorry for the late answer!

You have 2 options for this:
1) Include it in the OpenAPI document where you want the auth data to go (headers? body? post parameters?), and then it will be present in the generated code, and you can set it in a typesafe way.
2) Or if this is not an option, you could create a custom adapter, and influence the request in any way you like. Here is an example:
```ts
class YourCustomAdapter extends FetchClientAdapter {
  constructor(public readonly token: string) {
    // Or you can pass down config from the outside too, your choice
    super({ url: 'http://your.api.com' })
  }
  protected override getRequestInit(request: RawHttpRequest) {
    // requestInit here is untyped / any, so the client/server typings
    // don't bleed into each other, but in case you are extending
    // FetchClientAdapter, you can use all of this:
    // https://developer.mozilla.org/en-US/docs/Web/API/fetch#options
    const requestInit = super.getRequestInit(request)
    requestInit.credentials = 'foo'
    requestInit.headers = { ...(request.headers ?? {}), 'X-YOUR-HEADER':  this.token }
    return requestInit
  }
}

const yourSdk = new BookStoreSdkImpl(new YourCustomAdapter('token'))
// Use the SDK normally
```

Also a recommended read: https://oats-ts.github.io/docs/documentation/SdkPublish
